### PR TITLE
Handle vulnerability analyzer completion with Kafka Streams

### PR DESCRIPTION
**Pull request by nscuro**
### Description

This PR implements the tracking of vulnerability analyzer completion using Kafka Streams.

While the vulnerability analyzer application takes care of tracking scanner completion on a scan key (`<TOKEN>/<COMPONENT_UUID>`) basis, this PR enables the API server to track completion per scan token. 

Tracking it this way makes it possible to track completion across API server restarts, while not stressing the database.

### Addressed Issue

N/A

### Additional Details

* By observing the events sent to `dtrack.vuln-analysis.component`, we can deduce how many scan completion events we should expect for any given scan token (aka BOM upload token)
* Conversely, we can observe the scan completion events sent to `dtrack.vuln-analysis.result` that have the same scan token on them
* Through counting occurrences of those events, and comparing them, we can determine whether all expected events have been received

The analysis completion for each token is continuously updated as new events trickle in, and persisted into a RocksDB-backed key-value-store. This KV-store is then queried in the `/api/v1/bom/token/<TOKEN>` endpoint.

image

### Checklist

- [x] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/github-templates/CONTRIBUTING.md#pull-requests)
- [ ] ~This PR fixes a defect, and I have provided tests to verify that the fix is effective~
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [ ] ~This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)~
- [ ] ~This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly~

### CODE REVIEW

File: src/test/java/org/dependencytrack/event/kafka/KafkaStreamsTopologyTest.java [hunk 1]
```diff
@@ -164,7 +108,7 @@ public void testVulnResultIngestionNoFinding() {
         vulnAnalysisResultInputTopic.pipeInput(component.getUuid(), new VulnerabilityResult(null, AnalyzerIdentity.INTERNAL_ANALYZER, false));
 
         qm.getPersistenceManager().refresh(component);
-        assertThat(component.getLastVulnerabilityAnalysis()).isAfter(beforeAnalysis);
+        // assertThat(component.getLastVulnerabilityAnalysis()).isAfter(beforeAnalysis);
```
Comment (VithikaS) [inline src/test/java/org/dependencytrack/event/kafka/KafkaStreamsTopologyTest.java:111]: Do you need commented assertions enabled at some point or can they be just deleted ?

Comment (nscuro) [reply src/test/java/org/dependencytrack/event/kafka/KafkaStreamsTopologyTest.java]: It turned out that the entire class could be removed. The topology is now tested using a Redpanda test container, there is no need anymore to have a separate test class.

File: src/main/java/org/dependencytrack/tasks/BomUploadProcessingTask.java [hunk 2]
```diff
@@ -126,7 +124,7 @@ public void inform(final Event e) {
                     LOGGER.warn("The BOM uploaded is not in a supported format. Supported formats include CycloneDX XML and JSON");
                     return;
                 }
-                final Project copyOfProject = qm.detach(Project.class, qm.getObjectById(Project.class, project.getId()).getId());
+                // final Project copyOfProject = qm.detach(Project.class, qm.getObjectById(Project.class, project.getId()).getId());
```
Comment (VithikaS) [inline src/main/java/org/dependencytrack/tasks/BomUploadProcessingTask.java:127]: Super minor : just cosmetic ...we can remove it if this is not required? There is also another commented code just below in existing code base.

**Review by VithikaS — APPROVED** [review]
Added a very minor comment but looks great. Thanks for topology diagrams otherwise would be very difficult to follow what was happening

**Comment by nscuro** [comment]
I was able to resolve the test failures for #63. I think the best way forward would be to merge that PR first, then I can rebase this PR and fix any failures that are left over.

**Pull request merged**